src/topology/metric_space/hausdorff_distance.lean

```diff
@@ -0,0 +1,1181 @@
+      rw [← hy', ← inf_edist_image h],
+      exact mem_image_of_mem _ (mem_image_of_mem _ hy) }},
+  { ext b,
+    split,
+    { assume hb,
+      rcases (mem_image _ _ _ ).1 hb with ⟨y, ⟨hy, hy'⟩⟩,
+      rcases (mem_image _ _ _ ).1 hy with ⟨z, ⟨hz, hz'⟩⟩,
+      rw [← hy', ← hz', inf_edist_image h],
+      exact mem_image_of_mem _ hz },
+    { assume hb,
+      rcases (mem_image _ _ _ ).1 hb with ⟨y, ⟨hy, hy'⟩⟩,
+      rw [← hy', ← inf_edist_image h],
+      exact mem_image_of_mem _ (mem_image_of_mem _ hy) }}
+end
+
+/-- The Hausdorff distance is controlled by the diameter of the union -/
+lemma Hausdorff_edist_le_ediam (hs : s ≠ ∅) (ht : t ≠ ∅) : Hausdorff_edist s t ≤ diam (s ∪ t) :=
+begin
+  rcases ne_empty_iff_exists_mem.1 hs with ⟨x, xs⟩,
+  rcases ne_empty_iff_exists_mem.1 ht with ⟨y, yt⟩,
+  refine Hausdorff_edist_le_of_mem_edist _ _,
+  { exact λz hz, ⟨y, yt, edist_le_diam_of_mem (subset_union_left _ _ hz) (subset_union_right _ _ yt)⟩ },
+  { exact λz hz, ⟨x, xs, edist_le_diam_of_mem (subset_union_right _ _ hz) (subset_union_left _ _ xs)⟩ }
+end
+
+/-- The Hausdorff distance satisfies the triangular inequality -/
+lemma Hausdorff_edist_triangle : Hausdorff_edist s u ≤ Hausdorff_edist s t + Hausdorff_edist t u :=
+begin
+  change Sup ((λx, inf_edist x u) '' s) ⊔ Sup ((λx, inf_edist x s) '' u) ≤ Hausdorff_edist s t + Hausdorff_edist t u,
+  simp only [and_imp, set.mem_image, lattice.Sup_le_iff, exists_imp_distrib,
+             lattice.sup_le_iff, -mem_image, set.ball_image_iff],
+  split,
+  show ∀x ∈ s, inf_edist x u ≤ Hausdorff_edist s t + Hausdorff_edist t u, from λx xs, calc
+    inf_edist x u ≤ inf_edist x t + Hausdorff_edist t u : inf_edist_le_inf_edist_add_Hausdorff_edist
+    ... ≤ Hausdorff_edist s t + Hausdorff_edist t u :
+      add_le_add_right' (inf_edist_le_Hausdorff_edist_of_mem  xs),
+  show ∀x ∈ u, inf_edist x s ≤ Hausdorff_edist s t + Hausdorff_edist t u, from λx xu, calc
+    inf_edist x s ≤ inf_edist x t + Hausdorff_edist t s : inf_edist_le_inf_edist_add_Hausdorff_edist
+    ... ≤ Hausdorff_edist u t + Hausdorff_edist t s :
+      add_le_add_right' (inf_edist_le_Hausdorff_edist_of_mem xu)
+    ... = Hausdorff_edist s t + Hausdorff_edist t u : by simp [Hausdorff_edist_comm, add_comm]
+end
+
+/-- The Hausdorff edistance between a set and its closure vanishes -/
+@[simp] lemma Hausdorff_edist_self_closure : Hausdorff_edist s (closure s) = 0 :=
+begin
+  erw ← le_bot_iff,
+  simp only [Hausdorff_edist, inf_edist_closure, -le_zero_iff_eq, and_imp,
+    set.mem_image, lattice.Sup_le_iff, exists_imp_distrib, lattice.sup_le_iff,
+    set.ball_image_iff, ennreal.bot_eq_zero, -mem_image],
+  simp only [inf_edist_zero_of_mem, mem_closure_iff_inf_edist_zero, le_refl, and_self,
+             forall_true_iff] {contextual := tt}
+end
+
+/-- The Hausdorff edistance between sets or their closures is the same -/
+@[simp] lemma Hausdorff_edist_closure : Hausdorff_edist (closure s) (closure t) = Hausdorff_edist s t :=
+begin
```

Comment:
Perhaps prove `Hausdorff_edist (closure s) t = Hausdorff_edist s t` first, and then deduce this.

Reply:
Done